interoperator/docs/architecture/basic.md

```diff
@@ -0,0 +1,493 @@
+
+## Basic Control-flow
+
+### Catalog
+
+![Service Fabrik Inter-operator Basic Control-flow Catalog](images/basic-control-flow-catalog.png)
+
+#### Service and Plan registration
+
+The following steps are part of the landscape setup and a landscape administrator.
+This could be an actual person or could be an automated component in itself.
+
+1. Register `SFService` for the service.
+There would be one `SFService` instance per service in the landscape.
+
+It could be possible that a single Service Fabrik inter-operator serves multiple services in the same set of Kubernetes clusters. In such a case, there could be multiple `sfservices` registered for the same Service Fabrik inter-operator. But each of these `sfservices` would be for different individual services.
+
+2. Register `sfplans` for each plan supported by the service.
+As part of the away from t-shirt size approach to plans, it is recommended to minimize the number of plans per service. Ideally, that would be exactly one `SFPlan` per individual service.
+
+Updates to the services and plans can be done as simple updates to the corresponding `sfservices` and `sfplans`. Service and plans can be unregistered by simply deleting the corresponding `sfservices` and `sfplans`.
+
+TODO Backward compatibility existing instances must be handled by the individual service implemementations and the applications properly.
+
+#### Service Fabrik Broker Catalog Cache
+
+The Service Fabrik Broker watches for registered `sfservices` and `sfplans`. It reacts to registrations, updates and unregistrations and keeps an up-to-date representation of the information.
+
+#### Integraton with Service Manager
+
+1. An OSB client queries the [Service Manager](https://github.com/Peripli/service-manager) for a catalog of the available services via the `v2/catalog` request.
+1. The Service Manager forwards this call (via some possible intermediaries) to the Service Fabrik Broker. 
+1. The Service Fabrik Broker refers to its [internal up-to-date representation](#service-fabrik-broker-catalog-cache) and serves the catalog for the currently registered services.
+
+### Provision
+
+This section presumes that the `SFService` and `sfplans` are already registered as describe [above](#catalog).
+
+![Service Fabrik Inter-operator Basic Control-flow Provision](images/basic-control-flow-provision.png)
+
+#### Service Fabrik Broker
+
+1. An OSB client makes a `provision` call to the [Service Manager](https://github.com/Peripli/service-manager).
+1. The Service Manager forwards the call (perhaps via some intermediaries) to Service Fabrik Broker if the `provision` call was for a service and plan that was published by the Service Fabrik Broker.
+The Service Manager adds some relevant additional context into the request.
+1. The Service Fabrik Broker creates an `SFServiceInstance` capturing all the details passed in the the `provision` request from the Service Manager.
```

Comment:
I understand the need for SFServiceInstance CRD as this creates an intermediate OSB representation of the service. I have a suggestion which avoids this intermediate CRD. At individual service/plans we expect services to define the template (for both resource/status), so why not use that directly at the controller layers directly via the interoperator to create the service speicific CRD and also in lastOperation the status template to fetch the status back and return it? 
By doing so, we still achieve the end objective of not having services to adhere to OSB spec and additionally we completely eliminate the need for synching the resources/state between SFServiceInstance and service specific CRDs.
By doing so, Interoperator as a component will no longer be a controller but the functionality of it will be directly consumed in the OSB API controller layer (i.e. in SFBrokerAPIController layer directly).

Reply:
The SFServiceInstance resource is serving two different purposes.
1. Decouple scheduling (which cluster to provision the instance in) from the actual provisioning of the instance.
   * Granted that the scheduling problem is relevant only in a multi-cluster landscape. But I think single-cluster is workable only for the very short term (especially, for something like Postgres which is liable to have large number of instances). I am adding some more content to the cluster-landscape to explain the details of the comparison of the different cluster landscapes and the justification for the recommendation for the optimal landscape (multi-cluster).
Such decoupling is required in multi-cluster scenario just to be able to schedule among a pool of available clusters. At the very least, there is a need to keep two separate CRDs.
      1. One CRD in the cluster where SF (broker and the inter-operator) is running. This is the CRD on which the scheduler would work.
      1. The other CRD in the actual cluster where the instance is provisioned. This is the CRD on which the service’s operator would act on.
   * It is a different matter if such a separation is best done by having the SFServiceInstance instance in the cluster where SF is running and the service’s CRD in the cluster where it is to be provisioned or if we can replicate the same service’s CRD in both locations. This is addressed below.
   * Such a decoupling is required to provide a stable identity for the service instance even while moving the actual provisioning from one cluster to another (if such a need arises).
1. Capture the details about the OSB request verbatim before it is processed by the inter-operator.
   *  Granted that this is not of immediately obvious value. But there is some value in capturing the request details as spec verbatim, especially, in debugging, tracing but also in reconciliation or even in update/upgrade logic.
   * Especially, when we anyway need two CRDs as discussed above, I think having SFServiceInstance that captures additional information is useful when compared to just replicating the service’s CRD.
   * It also makes SF fully stateful regarding all its instances.
   * There are additional advantages of providing a uniform interface to the scheduler (potentially custom schedulers) but that is a very minor advantage which we can even work without.
 
What do you think?

Reply:
@jagadish-kb I have taken the liberty of cross posting from our email conversation here. I apologize if that is overstepping the bounds of decency.

> With regards to storing OSB request payload, one could always leverage annotations on service specific objects which are free flowing json.

Agreed. That was a weak argument. Only relevant in the presence of a different stronger argument as you agreed below.
 
> Decouple Scheduling is a strong reason, Yes I agree. I was of the impression that both the interoperator & services operator will be in same cluster. If they are in different cluster decoupling also helps in resiliency (Like the issue bosh-sf downtime today had issues with request processing in SF, which we had to address specifically). Also, apart from the other benefits you’ve mentioned additionally this decoupling could help us in streamlining some reporting requirements as well quite easily. My main concern was state synchronization, guess the above benefits now outweigh this concern.
Thanks for the clarification.

Thanks for both the counter-arguments as well as the eventual agreement and pointing our the additional benefits :-)